### ISSUE TITLE
Allow usage of `ring` instead of `aws-lc-rs` in `tokio-rustls`

### ISSUE THREAD

**Issue by Leandros** [hyperium/tonic#1943]
### Motivation

To allow for not linking to `aws-lc-rs` (specifically, removing any usages of `aws-lc-sys`), I'd like to have the ability to chose the used provider from rustls.

### Proposal

Tonic pulls in the `tokio-rustls` dependency, which defaults to use `aws-lc` as a cryptography primitive provider. However, `tokio-rustls` also ships with an alternative provider `ring`. To enable ring, the feature `ring` needs to be set (and the default features disabled). 

To allow this with tonic, tonic would have to expose a new feature that allows choosing `ring` over `aws-lc` and forwards the feature flags to `tokio-rustls`.

### Alternatives

Ideally, `cargo` would allow us to change features of dependencies of dependencies. However, at the current moment this isn't possible.

**Comment by mikemorris** [comment]
Has the direct ask here been resolved with https://github.com/hyperium/tonic/pull/2034?

Somewhat related however, it would be nice if we could configure Tonic to optionally use [alternative third-party providers](https://github.com/rustls/rustls?tab=readme-ov-file#third-party-providers).

**Comment by alexrudy** [comment]
Confirmed, #2034 and others solve this, and it will be supported in the next release. I don't know when that will be, but maybe @LucioFranco we consider another release, since it's been several months (Sept. 2024, IIUC)?

At that point, there will be two tls features, `tls-ring` and `tls-aws-lc-rs`, which will do what you want respectively.

**Comment by LucioFranco** [comment]
Do we still want to do this if https://rustsec.org/advisories/RUSTSEC-2025-0007.html just was posted a few days ago?

**Comment by LucioFranco** [comment]
As for a release, I've been out of the country (away from the desk for the last few weeks and won't be back for another week and a half plus jetlag) so I was planning on getting a release out then since nothing seemed extremely high priority (though if there is please do ping me).

**Comment by aumetra** [comment]
The advisory was almost immediately retracted since the rustls team stepped up for basic maintenance without new feature development.

I feel like it still offers some smaller advantages, mainly in the way it compiles pretty easily in comparison to aws-lc-rs, but that's preference.

**Comment by TheButlah** [comment]
Yep, for me the main benefit is the easier compilation, especially cross compilation.

**Comment by LucioFranco** [comment]
> The advisory was almost immediately retracted since the rustls team stepped up for basic maintenance without new feature development.

@aumetra do you have a link to this redaction I must have missed it but I am traveling (not for work) so maybe I missed it.

**Comment by youyuanwu** [comment]
Slightly off topic but I have seen more adoption of my project https://github.com/youyuanwu/tonic-tls to bypass ring recently.
I am willing to integrate/transfer it into tonic itself as a sub-crate, and I will keep maintaining it. Let me know.

**Comment by mikemorris** [comment]
Withdrawn advisory link for future reference https://rustsec.org/advisories/RUSTSEC-2025-0007.html

**Comment by LucioFranco** [comment]
Since this is fixed already in the master branch and a release will be going out this week, I will close this issue.

**Comment by LucioFranco** [comment]
Progress on the release can be tracked here https://github.com/hyperium/tonic/issues/2135

**Comment by mikemorris** [comment]
@LucioFranco should I open a new issue to track the possibility of making passthrough config more flexible as described in https://github.com/hyperium/tonic/issues/1943#issuecomment-2701371902?

**Comment by LucioFranco** [comment]
@mikemorris yeah that would be a good idea
